### PR TITLE
gh-103384: Generalize the regex pattern `BaseConfigurator.INDEX_PATTERN` to allow spaces and non-alphanumeric characters in keys.

### PR DESCRIPTION
**Pull request by dhuadaar**
The issue described a broken regex pattern in the logging module. 
Fixed the regex and added tests for the same.

* Issue: gh-103384

### CODE REVIEW

File: Lib/test/test_logging.py
Comment (hugovk) [inline Lib/test/test_logging.py]: Typo:
```suggestion
            'adict': {'d': 'e', 'f': 3 , 'alpha numeric 1 with spaces' : 5},
```

File: Lib/test/test_logging.py
Comment (hugovk) [inline Lib/test/test_logging.py]:
```suggestion
        self.assertEqual(bc.convert('cfg://adict[alpha numeric 1 with spaces]'), 5)
```

File: Lib/logging/config.py
Comment (tomasr8) [inline Lib/logging/config.py]: I wonder if we want to allow something like `nested[]]` where the inner `]` is interpreted as a key

**Comment by tomasr8** [comment]
I think a test with some special characters would also be useful :)

**Comment by dhuadaar** [comment]
> I think a test with some special characters would also be useful :)

Makes sense.

Thanks for the suggestions. I will add some of those combinations.

File: Lib/logging/config.py
Comment (dhuadaar) [reply Lib/logging/config.py]: I did not find a lot of clarity in the documentation around the charset allowed. However, I don't see a reason why not.
Let's wait for a few more people to review.

In case you are aware where the specs are, please point me to the same.

Comment (tomasr8) [reply Lib/logging/config.py]: I haven't found anything besides what is linked in the issue, I guess it depends on how sophisticated we want it to be..

**Comment by tomasr8** [comment]
I'd also add tests for some edge cases like the empty string (if we want to allow that) and multiple nesting levels e.g. `[x][y]` or even `[x].y[z]` to make sure the regex is working as expected.

File: Lib/test/test_logging.py
Comment (arhadthedev) [inline Lib/test/test_logging.py]: Should it be `alpha numeric` instead of `alpha nmmeric`?

Comment (dhuadaar) [reply Lib/test/test_logging.py]: Yes.
I have addressed this. 
Sorry about it.

**Comment by dhuadaar** [comment]
> I'd also add tests for some edge cases like the empty string (if we want to allow that) and multiple nesting levels e.g. `[x][y]` or even `[x].y[z]` to make sure the regex is working as expected.

I have added a new nested dictionary to the existing template we use in the test. Added new assertions specifically for the nested cases. 

Do you suggest making the cases like the following to the part of the new key added?

`
'nest2': ['k', ['l', 'm'], 'n'],
`

File: Lib/logging/config.py
Comment (vsajip) [reply Lib/logging/config.py]: I don't think we need to support anything beyond simple keys (e.g. not `nested[]]` to actually mean nesting as suggested above) but there should be tests with `]` in the key to make sure the correct value is used. For example tests with keys `abc ] def` and `abc]` to ensure the correct key is computed. Likewise, tests with leading and trailing spaces. Note that allowing leading and trailing spaces, as the proposed regex does, could lead to annoyances because of accidentally inserted spaces.

File: Misc/NEWS.d/next/Library/2023-04-09-05-30-41.gh-issue-103384.zAV7iB.rst
Comment (vsajip) [inline Misc/NEWS.d/next/Library/2023-04-09-05-30-41.gh-issue-103384.zAV7iB.rst]:
```suggestion
Generalize the regex pattern `BaseConfigurator.INDEX_PATTERN` to allow spaces and non-alphanumeric characters in keys.
```

**Review by vsajip — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: Lib/logging/config.py
Comment (vsajip) [inline Lib/logging/config.py]: Maybe allowing leading spaces inside the `[` would be confusing?

**Comment by dhuadaar** [comment]
@vsajip 

Thanks. I will fix these soon.

File: Lib/logging/config.py
Comment (dhuadaar) [reply Lib/logging/config.py]: I believe yes.
However, given there is no clear specification so far, we can leave the specifics to the end user.

File: Misc/NEWS.d/next/Library/2023-04-09-05-30-41.gh-issue-103384.zAV7iB.rst
Comment (arhadthedev) [inline Misc/NEWS.d/next/Library/2023-04-09-05-30-41.gh-issue-103384.zAV7iB.rst]:
```suggestion
Generalize the regex pattern ``BaseConfigurator.INDEX_PATTERN`` to allow spaces and non-alphanumeric characters in keys. Patch by Peeyush Aggarwal.
```

File: Lib/logging/config.py
Comment (dhuadaar) [reply Lib/logging/config.py]: Just seeking a clarification here. 

If we have a key `abc ] def` and we try to use it in the cfg notation as follows:

```
cfg://adict[abc ] def ]
```
Shouldn't it technically evaluate as a key `abc `?
Am I missing something here?

Going by this, I think although non alpha numeric keys are allowed, but allowing `[` and `]` would pose problem to the larger syntax convention being used.

Comment (dhuadaar) [reply Lib/logging/config.py]: here is a sample run to explain what I suggested

```
import logging
import logging.handlers
import logging.config

sample = {"a" : 1, "a]b" : 4, "a][b" : 12, a" : 3, "a b" : 5}
config = {'adict' : sample}

bc = logging.config.BaseConfigurator(config)

>>> bc.convert('cfg://adict[a]')
1
>>> 

>>> bc.convert('cfg://adict[ a]')
1

>>> bc.convert('cfg://adict[a b]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 461, in convert
    value = converter(suffix)
            ^^^^^^^^^^^^^^^^^
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 431, in cfg_convert
    raise ValueError('Unable to convert '
ValueError: Unable to convert 'adict[a b]' at '[a b]'


>>> bc.convert('cfg://adict[a][b]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 461, in convert
    value = converter(suffix)
            ^^^^^^^^^^^^^^^^^
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 421, in cfg_convert
    d = d[idx]
        ~^^^^^
TypeError: 'int' object is not subscriptable


>>> bc.convert('cfg://adict[a]b]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 461, in convert
    value = converter(suffix)
            ^^^^^^^^^^^^^^^^^
  File "/Users/peeyush/.pyenv/versions/3.11.1/lib/python3.11/logging/config.py", line 431, in cfg_convert
    raise ValueError('Unable to convert '
ValueError: Unable to convert 'adict[a]b]' at 'b]'

```

Comment (vsajip) [reply Lib/logging/config.py]: Thinking about this, we should probably disallow keys with `]` at the very least, and perhaps `[` too because it will just make any actual usage with these quite confusing when eye-balling. Or perhaps only allow alphanumerics and spaces, as that will be consistent with the documentation as it is now.

Comment (dhuadaar) [reply Lib/logging/config.py]: Makes sense to me.

The document reads

```
The latter form only needs to be used if the key contains spaces or non-alphanumeric characters.
```

So special characters are indeed allowed. I think the regex pattern which is configured right now is okay. 

It will match for the very first `]` as expected.
The user would intuitively understand this constraint. It will anyways raise an exception. I will add tests for that.

However, we can improve the [documentation](https://raw.githubusercontent.com/python/cpython/main/Doc/library/logging.config.rst) as well. 

We can add - 

```
Please note that the key cannot contain `[` and `]`.
```

Update:

- Added a few test cases for failing scenarios.
- Modified the regex to now all `[` and `]` within key.

Comment (vsajip) [reply Lib/logging/config.py]: > We can add

Sure, feel free to do that in this PR as well.

**Comment by dhuadaar** [comment]
@vsajip 

I have made the requested changes; please review again.

**Comment by dhuadaar** [comment]
@vsajip 
Could you review this PR? I have made the changes requested.

**Comment by vsajip** [comment]
Sorry I've not had time to look at this. Hope to get to it soon.

**Review by vsajip — APPROVED** [review]

**Comment by erlend-aasland** [comment]
I'll leave it for the code owner @vsajip to land this :) Thanks for your contribution, @dhuadaar!

**Comment by erlend-aasland** [comment]
Should this be backported?

**Comment by vsajip** [comment]
Perhaps - it's arguable whether it's an enhancement or could have been considered a documentation bug.

**Comment by erlend-aasland** [comment]
Ok; if no backport is needed, we should close the linked issue.

**Pull request merged**
